// @electron-forge/template-typescript-webpack/dist/TypeScriptWebpackTemplate.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = void 0;
var _asyncOra = require("@electron-forge/async-ora");
var _templateBase = require("@electron-forge/template-base");
var _fsExtra = _interopRequireDefault(require("fs-extra"));
var _path = _interopRequireDefault(require("path"));
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
class TypeScriptWebpackTemplate extends _templateBase.BaseTemplate {
    async initializeTemplate(directory, options) {
        await super.initializeTemplate(directory, options);
        await (0, _asyncOra).asyncOra('Setting up Forge configuration', async ()=>{
            const packageJSONPath = _path.default.resolve(directory, 'package.json');
            const packageJSON = await _fsExtra.default.readJson(packageJSONPath);
            packageJSON.main = '.webpack/main';
            packageJSON.config.forge.plugins = packageJSON.config.forge.plugins || [];
            packageJSON.config.forge.plugins.push([
                '@electron-forge/plugin-webpack',
                {
                    mainConfig: './webpack.main.config.js',
                    renderer: {
                        config: './webpack.renderer.config.js',
                        entryPoints: [
                            {
                                html: './src/index.html',
                                js: './src/renderer.ts',
                                name: 'main_window',
                                preload: {
                                    js: './src/preload.ts'
                                }
                            }, 
                        ]
                    }
                }, 
            ]);
            // Configure scripts for TS template
            packageJSON.scripts.lint = 'eslint --ext .ts,.tsx .';
            await _fsExtra.default.writeJson(packageJSONPath, packageJSON, {
                spaces: 2
            });
        });
        await (0, _asyncOra).asyncOra('Setting up TypeScript configuration', async ()=>{
            const filePath = (fileName)=>_path.default.join(directory, 'src', fileName)
            ;
            // Copy Webpack files
            await this.copyTemplateFile(directory, 'webpack.main.config.js');
            await this.copyTemplateFile(directory, 'webpack.renderer.config.js');
            await this.copyTemplateFile(directory, 'webpack.rules.js');
            await this.copyTemplateFile(directory, 'webpack.plugins.js');
            await this.updateFileByLine(_path.default.resolve(directory, 'src', 'index.html'), (line)=>{
                if (line.includes('link rel="stylesheet"')) return '';
                return line;
            });
            // Copy tsconfig with a small set of presets
            await this.copyTemplateFile(directory, 'tsconfig.json');
            // Copy eslint config with recommended settings
            await this.copyTemplateFile(directory, '.eslintrc.json');
            // Remove index.js and replace with index.ts
            await _fsExtra.default.remove(filePath('index.js'));
            await this.copyTemplateFile(_path.default.join(directory, 'src'), 'index.ts');
            await this.copyTemplateFile(_path.default.join(directory, 'src'), 'renderer.ts');
            // Remove preload.js and replace with preload.ts
            await _fsExtra.default.remove(filePath('preload.js'));
            await this.copyTemplateFile(_path.default.join(directory, 'src'), 'preload.ts');
        });
    }
    constructor(...args){
        super(...args);
        this.templateDir = _path.default.resolve(__dirname, '..', 'tmpl');
    }
}
var _default = new TypeScriptWebpackTemplate();
exports.default = _default;
