// @xmcl/nbt/zlib/index.browser.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var pako = _interopDefault(require('pako'));

function gzip(buffer) { return Promise.resolve(pako.gzip(buffer)); }
function gzipSync(buffer) { return pako.gzip(buffer); }
function ungzip(buffer) { return Promise.resolve(pako.ungzip(buffer)); }
function gunzipSync(buffer) { return pako.ungzip(buffer); }
function inflate(buffer) { return Promise.resolve(pako.inflate(buffer)); }
function deflate(buffer) { return Promise.resolve(pako.deflate(buffer)); }
function inflateSync(buffer) { return pako.inflate(buffer); }
function deflateSync(buffer) { return pako.deflate(buffer); }

exports.deflate = deflate;
exports.deflateSync = deflateSync;
exports.gunzipSync = gunzipSync;
exports.gzip = gzip;
exports.gzipSync = gzipSync;
exports.inflate = inflate;
exports.inflateSync = inflateSync;
exports.ungzip = ungzip;
